### PR TITLE
optimizaiton softmax entropy

### PR DESCRIPTION
**Pull request by xingfeng01**
### PR types
Performance optimization

### PR changes
OPs

### Describe
Op softmax_with_cross_entropy optimization. Main idea is described in code comments.

### CODE REVIEW

**Comment by lanxianghit** [comment]
description为空，这是开发规范不允许的，请按照要求说清本PR类型、主要修改、以及相关能够使其他人清楚pr目的的信息。

File: paddle/fluid/operators/softmax_cudnn_op.cu [hunk 1]
```diff
@@ -51,13 +51,13 @@ using Tensor = framework::Tensor;
...
-  }
-  return size;
-}
+// static inline int SizeOutAxis(const int axis, DDim dims) {
+//   int size = 1;
+//   for (int i = axis + 1; i < dims.size(); i++) {
+//     size *= dims[i];
+//   }
+//   return size;
+// }
```
Comment (lanxianghit) [inline paddle/fluid/operators/softmax_cudnn_op.cu]: 为什么使用注释？如果是确定不需要的代码，应该直接删除，如果确实需要注释，也需要写明原因。

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 2]
```diff
@@ -70,61 +110,63 @@ __global__ void SoftCrossEntropyGradientKernel(T* logit_grad,
...
-  CUDA_KERNEL_LOOP(index, num) {
-    int idx_n = index / d;
-    int idx_remain = index % remain;
-    int idx_lbl = idx_n * remain + idx_remain;
-    int k = (index % d) / remain;
-    if (labels[idx_lbl] == ignore_index || labels[idx_lbl] != k) {
-      logit_grad[index] = static_cast<T>(0.);
-    } else {
-      logit_grad[index] *= loss_grad[idx_lbl];
+__global__ void CrossEntropySoftKernel(T* loss, const T* softmax,
```
Comment (lanxianghit) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: 代码整体结构变动比较大，有较多增删，最好在关键位置加一些简要的注释，说明目的，也便于后续其他人维护

**Review by lanxianghit — CHANGES_REQUESTED** [review]

**Comment by lanxianghit** [comment]
代码整体改动较多，需要看一下是否需要增加单测

**Comment by xingfeng01** [comment]
性能变优，GPU运行时间减少：
softmax_with_cross_entropy_0: -36% (forward), -30% (backward)
softmax_with_cross_enttropy_1: - 89% (forward), -79% (backward)

File: paddle/fluid/operators/softmax_cudnn_op.cu (hunk 1 above)
Comment (xingfeng01) [reply paddle/fluid/operators/softmax_cudnn_op.cu]: 注释已删除。

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu (hunk 2 above)
Comment (xingfeng01) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: 注释已经加如代码中，解释函数功能和主要的实现。

**Comment by xingfeng01** [comment]
Failure CI PR-CI-APPROVAL 显示 Using ShareDataWith or ShareBufferWith is not recommended. 这部分代码非新加入。

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 3]
```diff
@@ -15,829 +15,1007 @@ limitations under the License. */
...
-  int64_t end_idx = (idx_n + 1) * d;
-
-  int64_t step = BlockDim * remain;
-  T cur_max = logits_data[beg_idx];
-  beg_idx += step;
-  while (beg_idx < end_idx) {
-    if (cur_max < logits_data[beg_idx]) {
-      cur_max = logits_data[beg_idx];
+template <typename T, typename VecT>
+__global__ void CrossEntropySoftLabel(T* loss, const T* softmax,
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu:159]: 这2个函数的差别是什么？

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 4]
```diff
@@ -15,829 +15,1007 @@ limitations under the License. */
...
+        }
+      }
+
+      int idx = threadIdx.x + it * kWarpSize;
+      if (idx < idx_max_v) {
+        softmax_v[idx] = tmpv;
+      }
     }
   }
+}
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu:554]: 这个函数有140行了，太长了。

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 5]
```diff
@@ -15,829 +15,1007 @@ limitations under the License. */
...
       for_range(HardLabelSoftmaxWithCrossEntropyFunctorWithIgnoreIdx<T>(       \
-          labels_data, loss_data, softmax_data, d, axis_dim, ignore_idx));     \
+          labels_data, loss_data, softmax_data, dim * d, dim, ignore_idx));    \
     } else {                                                                   \
       for_range(HardLabelSoftmaxWithCrossEntropyFunctor<T>(                    \
-          labels_data, loss_data, softmax_data, d, axis_dim, ignore_idx));     \
+          labels_data, loss_data, softmax_data, dim * d, dim, ignore_idx));    \
     }                                                                          \
   } break
+#endif
```
Comment (Xreki) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: rocm是不是不走下面这个版本？这里应该用`#else`？

**Review by Xreki — COMMENTED** [review]
代码改动太多，有2个小建议：

- 对相同功能的代码块进行封装。
- 代码拆成多个PR

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu (hunk 5 above)
Comment (xingfeng01) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: rocm相关问题已修改。

**Comment by xingfeng01** [comment]
此PR不再需要。已合入PR 31821 实现同样功能。
